### PR TITLE
Narayana recovery manager service

### LINKED ISSUE #26160: Improve recovery manager in quarkus-narayana-jta extension
### Description

Currently it only produces `RecoveryManagerService` and no any code for starting and stoping. I think recovery manager is core component of Narayana and it needs to be to reconcile transactions that fail.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by zhfeng**
Fix https://github.com/quarkusio/quarkus/issues/26160

### CODE REVIEW

**Comment by mmusgrov** [comment]
I like this PR, thanks for adding it Amos.

I need some time to study it though. For example we need be able to test it to verify recovery. I was looking through the narayana ones [1], some of them require byteman others don't so we need to figure out which functionality we can test in Quarkus without byteman.

I would also like to compare and contrast it with how we configure recovery in WildFly [2]

[1] https://github.com/jbosstm/narayana/tree/master/ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery
[2] https://github.com/wildfly/wildfly/blob/main/transactions/src/main/java/org/jboss/as/txn/service/ArjunaRecoveryManagerService.java#L91

**Comment by zhfeng** [comment]
And it looks like there is an issue for byteman extension https://github.com/quarkusio/quarkus/issues/2263

**Comment by mmusgrov** [comment]
Thanks for the last change @zhfeng 
Do you know who will be adding tests for this PR - is it something you need to do or is it something for the narayana team?

**Comment by zhfeng** [comment]
Hi @mmusgrov , I think we can always work together to add the tests. It could leverage https://github.com/jbosstm/quickstart/tree/master/spring/narayana-spring-boot to do the recovery test manually without byteman. But I did not find a way to restart a quarkus application after crashing just like we did in Arqullian container, e.g. https://github.com/jbosstm/narayana/blob/master/XTS/localjunit/crash-recovery-tests/src/test/java/com/arjuna/qa/junit/BaseCrashTest.java#L192

**Comment by zhfeng** [comment]
Hi @mmusgrov ,

I just create a simple project [1] to demonstrates the crash recovery. Please can you verify it on your side? and you need to rebuild the quarkus with the changes in this PR.

Thanks,

[1] https://github.com/zhfeng/quarkus-narayana-crash-recovery

**Comment by zhfeng** [comment]
@mmusgrov can you review these changes? I'd test it manually both work in JVM and native mode.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java [hunk 1]
```diff
@@ -30,4 +31,25 @@ public final class TransactionManagerConfiguration {
      */
     @ConfigItem(defaultValue = "ObjectStore")
     public String objectStoreDirectory;
+
+    /**
+     * The list of recovery modules
+     */
+    @ConfigItem(defaultValue = "com.arjuna.ats.internal.arjuna.recovery.AtomicActionRecoveryModule," +
```
Comment (mmusgrov) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java:38]: +1 I agree, I think this is all we need for recovering XA resources.

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java [hunk 2]
```diff
@@ -30,4 +31,25 @@ public final class TransactionManagerConfiguration {
...
+    @ConfigItem(defaultValue = "com.arjuna.ats.internal.arjuna.recovery.ExpiredTransactionStatusManagerScanner")
+    public List<String> expiryScanners;
+
+    /**
+     * The list of orphan filters
+     */
+    @ConfigItem(defaultValue = "com.arjuna.ats.internal.jta.recovery.arjunacore.JTATransactionLogXAResourceOrphanFilter," +
+            "com.arjuna.ats.internal.jta.recovery.arjunacore.JTANodeNameXAResourceOrphanFilter," +
+            "com.arjuna.ats.internal.jta.recovery.arjunacore.JTAActionStatusServiceXAResourceOrphanFilter")
+    public List<String> xaResourceOrphanFilters;
```
Comment (mmusgrov) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java:54]: +1 I agree, I think this is all we need for recovering orphaned XA branches.

**Review by mmusgrov — APPROVED** [review]
Thanks @zhfeng for this vital contribution.

Thanks also for providing your git branch for the [manual recovery test](https://github.com/zhfeng/quarkus-narayana-crash-recovery) which I have verified using your README. If we can get some input from the quarkus team to help us test crash recovery we should add some tests (basically we need to be able to crash and restart the test container and then verify that transactions are recovered). Amos did you approach anyone about this, @gastaldi any idea who we can ping to get some help with automating crash recovery testing?

File: extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java [hunk 3]
```diff
@@ -30,4 +31,25 @@ public final class TransactionManagerConfiguration {
      */
     @ConfigItem(defaultValue = "ObjectStore")
     public String objectStoreDirectory;
+
+    /**
+     * The list of recovery modules
+     */
+    @ConfigItem(defaultValue = "com.arjuna.ats.internal.arjuna.recovery.AtomicActionRecoveryModule," +
+            "com.arjuna.ats.internal.jta.recovery.arjunacore.XARecoveryModule")
+    public List<String> recoveryModules;
```
Comment (mmusgrov) [inline extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java:40]: +1 I agree, I think this is all we need for recovering XA resources.

**Comment by zhfeng** [comment]
Thanks @gastaldi I open https://github.com/quarkusio/quarkus/issues/27911 to add crash recovery support in testing. Anyway, can we get this fix into the coming 2.13 release since we can verify the recovery manually at least?

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
@zhfeng sure, I just merged because @mmusgrov approved 😉 
Thanks for the PR!

**Comment by mmusgrov** [comment]
> > any idea who we can ping to get some help with automating crash recovery testing?
> 
> @mmusgrov It sounds like a feature request to our testing harness library, mind opening a new issue with the expected behavior for this use case?

I have created #27958 to track the feature request.

**Comment by gsmet** [comment]
@mmusgrov @zhfeng we had reports of people complaining that Quarkus was now creating `ObjectStore` directories in the root modules of applications under test.

This is definitely not something we want.
I'm also a bit unsure of if this new feature should be enabled by default or if it should be protected by a config boolean.

@Sanne what's your take on this?

From my side, if we don't have a fix for the `ObjectStore` directories being created where they shouldn't, I will have to revert this patch for 2.13 as that's really not something we can afford (especially since there's a good chance users will commit the ObjectStores to their repositories without noticing).

I release 2.13.0.Final on Wednesday so we need a fix in by then or I will have to revert this patch in 2.13 (and we can keep it in main but we will need a fix before 2.14 is out).

**Comment by zhfeng** [comment]
@gsmet I think it can be fixed by set the default value of `objectStoreDirectory` to `target/ObjectStore`. And I think the crash recovery capability is very important to keep the data consistence during transactions. We could introduce an option `enableRecovery` but default valude should be `true`, and if users set it to `false`, it should print a WARN message. @mmusgrov WDYT?

**Comment by zhfeng** [comment]
Fix for `ObjectStore` https://github.com/quarkusio/quarkus/pull/28014

**Comment by zhfeng** [comment]
I just wonder why we should filter out the recovery starting messages in #27987 ?

**Comment by gsmet** [comment]
@zhfeng we work hard to keep the Quarkus startup log free from any noise. If we had every service of every extension logging two lines for the startup, we would end up with dozens of lines. We don't want that in Quarkus. We log if something goes wrong.

**Comment by gsmet** [comment]
@zhfeng I commented in your PR. I don't think that it's that simple. Otherwise I would have provided a PR myself :).

**Comment by gsmet** [comment]
@zhfeng is this recovery service needed even if XA is not enabled?

**Comment by zhfeng** [comment]
https://github.com/quarkusio/quarkus/pull/28025 to introduce `enableXARecovery`

**Comment by zhfeng** [comment]
I update the `objectStoreDirecotry` to use `OutputTargetBuildItem` as default value, see https://github.com/quarkusio/quarkus/pull/28025/commits/ca64491c16efcca28103bf2450a30d4563913538

**Pull request merged**